File: commands/stats/clan_24h.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone, timedelta
from collections import Counter
import re
from database import db
class Clan24h(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
    @app_commands.command(name='clan_24h', description='Clan statistics in last 24 hours')
    async def clan_24h(self, interaction: discord.Interaction, clan: str):
        try:
            await interaction.response.defer()
            clan_upper = clan.upper().strip()
            cutoff_time = datetime.now(timezone.utc) - timedelta(hours=24)
            query = {
                'clan_name': {'$regex': f'^{re.escape(clan_upper)}$', '$options': 'i'},
                'timestamp': {'$gte': cutoff_time}
            }
            
            pipeline = [
                {'$match': query},
                {'$facet': {
                    'stats': [{'$group': {
                        '_id': None,
                        'total': {'$sum': 1},
                        'contests': {'$sum': {'$cond': ['$contest', 1, 0]}},
                        'points': {'$sum': {'$cond': ['$contest', {'$multiply': ['$player_count', 2]}, '$player_count']}}
                    }}],
                    'maps': [{'$group': {'_id': '$map', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 3}],
                    'players': [{'$unwind': '$clan_winners'}, {'$group': {'_id': '$clan_winners', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 3}],
                    'recent': [{'$sort': {'timestamp': -1}}, {'$limit': 3}, {'$project': {'map': 1}}]
                }}
            ]
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            if not result or not result[0].get('stats'):
                await interaction.followup.send(embed=discord.Embed(
                    title=f"[{clan_upper}] Last 24 Hours", 
                    description="No wins in last 24 hours.", 
                    color=0xff0000
                ))
                return
            data = result[0]
            stats = data['stats'][0] if data['stats'] else {}
            total_wins = stats.get('total', 0)
            contest_wins = stats.get('contests', 0)
            total_points = stats.get('points', 0)
            top_maps = [(m['_id'], m['count']) for m in data.get('maps', [])]
            top_players = [(p['_id'], p['count']) for p in data.get('players', [])]
            recent = data.get('recent', [])
            
            embed = discord.Embed(
                title=f"[{clan_upper}] Last 24 Hours",
                description="Activity in the past day",
                color=0x2b2d31
            )
            
            stats_text = f"Wins: **{total_wins}**\n"
            stats_text += f"Contests: **{contest_wins}**\n"
            stats_text += f"Points: **{total_points}**\n"
            stats_text += f"Avg/Hour: **{total_wins/24:.1f}**"
            embed.add_field(name="📊 Stats", value=stats_text, inline=False)
            
            if top_maps:
                map_text = '\n'.join([f"{i+1}. {name} - {count}" for i, (name, count) in enumerate(top_maps)])
                embed.add_field(name="🗺️ Top Maps", value=map_text, inline=True)
            
            if top_players:
                player_text = '\n'.join([f"{i+1}. {name} - {count}" for i, (name, count) in enumerate(top_players)])
                embed.add_field(name="👥 Top Players", value=player_text, inline=True)
            
            recent_text = '\n'.join([f"{w.get('map', 'Unknown')}" for w in recent])
            embed.add_field(name="🕐 Recent", value=recent_text, inline=False)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in clan_24h: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to fetch 24h stats.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(Clan24h(bot))

File: commands/stats/clan_24h.js
// https://github.com/viktorexe/terristats-discord-bot
// /clan_24h - clan activity summary for the last 24 hours.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { exactMatch, gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('clan_24h')
    .setDescription('Clan statistics in last 24 hours')
    .addStringOption((o) =>
      o.setName('clan').setDescription('Clan name').setRequired(true),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const clan = interaction.options.getString('clan');
      const clanUpper = clan.toUpperCase().trim();
      const cutoff = new Date(Date.now() - 24 * 3600 * 1000);
      const query = { clan_name: exactMatch(clanUpper), timestamp: { $gte: cutoff } };

      const pipeline = [
        { $match: query },
        {
          $facet: {
            stats: [
              {
                $group: {
                  _id: null,
                  total: { $sum: 1 },
                  contests: { $sum: { $cond: ['$contest', 1, 0] } },
                  points: {
                    $sum: {
                      $cond: [
                        '$contest',
                        { $multiply: ['$player_count', 2] },
                        '$player_count',
                      ],
                    },
                  },
                },
              },
            ],
            maps: [
              { $group: { _id: '$map', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 3 },
            ],
            players: [
              { $unwind: '$clan_winners' },
              { $group: { _id: '$clan_winners', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 3 },
            ],
            recent: [
              { $sort: { timestamp: -1 } },
              { $limit: 3 },
              { $project: { map: 1 } },
            ],
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      if (!result.length || !result[0].stats.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`[${clanUpper}] Last 24 Hours`)
              .setDescription('No wins in last 24 hours.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const data = result[0];
      const stats = data.stats[0] || {};
      const totalWins = stats.total || 0;
      const contestWins = stats.contests || 0;
      const totalPoints = stats.points || 0;
      const topMaps = data.maps.map((m) => [m._id, m.count]);
      const topPlayers = data.players.map((p) => [p._id, p.count]);
      const recent = data.recent || [];

      const embed = new EmbedBuilder()
        .setTitle(`[${clanUpper}] Last 24 Hours`)
        .setDescription('Activity in the past day')
        .setColor(0x2b2d31);

      let statsText = `Wins: **${totalWins}**\n`;
      statsText += `Contests: **${contestWins}**\n`;
      statsText += `Points: **${totalPoints}**\n`;
      statsText += `Avg/Hour: **${(totalWins / 24).toFixed(1)}**`;
      embed.addFields({ name: '📊 Stats', value: statsText, inline: false });

      if (topMaps.length) {
        embed.addFields({
          name: '🗺️ Top Maps',
          value: topMaps.map(([n, c], i) => `${i + 1}. ${n} - ${c}`).join('\n'),
          inline: true,
        });
      }
      if (topPlayers.length) {
        embed.addFields({
          name: '👥 Top Players',
          value: topPlayers.map(([n, c], i) => `${i + 1}. ${n} - ${c}`).join('\n'),
          inline: true,
        });
      }

      embed.addFields({
        name: '🕐 Recent',
        value: recent.map((w) => `${w.map || 'Unknown'}`).join('\n'),
        inline: false,
      });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in clan_24h: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to fetch 24h stats.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
